src/test/java/org/hisp/dhis/rules/RuleEngineVariableNameTests.java

```diff
@@ -0,0 +1,545 @@
+            RuleEvent.Status.ACTIVE, new Date(), new Date(), "", null, Arrays.asList(
+                RuleDataValue.create( new Date(), "test_program_stage", "test_data_element_one", "condition" ) ), "" );
+
+        ruleEngineBuilder.events( Arrays.asList( ruleEvent2, ruleEvent3 ) );
+
+        List<RuleEffect> ruleEffects = ruleEngineBuilder.build().evaluate( ruleEvent ).call();
+
+        assertThat( ruleEffects.size() ).isEqualTo( 1 );
+        assertThat( ruleEffects.get( 0 ).ruleAction() ).isEqualTo( ruleAction );
+        assertTrue( ruleEffects.get( 0 ).data().equals( "2" ) );
+    }
+
+    @Test
+    public void evaluateD2CountIfVariableName()
+        throws Exception
+    {
+        RuleAction ruleAction = RuleActionDisplayKeyValuePair.createForFeedback(
+            "test_action_content", "d2:count(#{"+VARIABLE_NAME+"})" );
+        RuleVariable ruleVariableOne = RuleVariableNewestEvent.create(
+            VARIABLE_NAME, "test_data_element_one", RuleValueType.TEXT );
+
+        RuleVariable ruleVariableTwo = RuleVariableNewestEvent.create(
+            "test_var_two", "test_data_element_two", RuleValueType.TEXT );
+
+        Rule rule = Rule.create( null, null, "true", Arrays.asList( ruleAction ), "" );
+
+        RuleEngine.Builder ruleEngineBuilder = getRuleEngineBuilder( rule, Arrays.asList( ruleVariableOne ) );
+
+        RuleEvent ruleEvent = RuleEvent.create( "test_event", "test_program_stage",
+            RuleEvent.Status.ACTIVE, new Date(), new Date(), "", null, Arrays.asList(
+                RuleDataValue.create( new Date(), "test_program_stage", "test_data_element_one", "condition" ) ), "" );
+        RuleEvent ruleEvent2 = RuleEvent.create( "test_event2", "test_program_stage2",
+            RuleEvent.Status.ACTIVE, new Date(), new Date(), "", null, Arrays.asList(
+                RuleDataValue.create( new Date(), "test_program_stage", "test_data_element_one", "condition2" ) ), "" );
+        RuleEvent ruleEvent3 = RuleEvent.create( "test_event3", "test_program_stage3",
+            RuleEvent.Status.ACTIVE, new Date(), new Date(), "", null, Arrays.asList(
+                RuleDataValue.create( new Date(), "test_program_stage", "test_data_element_one", "condition" ) ), "" );
+
+        RuleEvent ruleEvent4 = RuleEvent.create( "test_event3", "test_program_stage3",
+            RuleEvent.Status.ACTIVE, new Date(), new Date(), "", null, Arrays.asList(
+                RuleDataValue.create( new Date(), "test_program_stage", "test_data_element_two", "condition" ) ), "" );
+
+        ruleEngineBuilder.events( Arrays.asList( ruleEvent2, ruleEvent3, ruleEvent4 ) );
+
+        List<RuleEffect> ruleEffects = ruleEngineBuilder.build().evaluate( ruleEvent ).call();
+
+        assertThat( ruleEffects.size() ).isEqualTo( 1 );
+        assertThat( ruleEffects.get( 0 ).ruleAction() ).isEqualTo( ruleAction );
+        assertEquals( "3", ruleEffects.get( 0 ).data() );
```

Comment:
This is correct, but I would expect the ruleEvent also to be passed in, as it is part of the context as well - and this would push the count up to 4.

Reply:
You are right but this is counting the test_data_element_one events.
If you look ruleEvent4 is linked to test_data_element_two, this is why it is not counted.

The count is 2 from the events + 1 from the current event